lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1621,8 +1621,9 @@ def _change_datatype(
         if isinstance(item, HistoryDatasetAssociation):
             self.hda_manager.ensure_can_change_datatype(item)
             self.hda_manager.ensure_can_set_metadata(item)
-            item.dataset.state = item.dataset.states.SETTING_METADATA
-            trans.sa_session.flush()
+            if not item.has_deferred_data:
+                item.dataset.state = item.dataset.states.SETTING_METADATA
+                trans.sa_session.flush()
```

Comment:
```suggestion
            item.dataset.state = item.dataset.states.SETTING_METADATA
            trans.sa_session.flush()
```
I would still set the state to SETTING_METADATA, so no-one can start a job with this dataset as input. Thinking about it, maybe `ensure_can_set_metadata` (or just `ok_to_edit_metadata` without the exception) should go into the task as well (See the second paragraph in https://docs.celeryq.dev/en/latest/userguide/tasks.html#state, `“asserting the world is the responsibility of the task”.`).

Reply:
Ok, I've done some changes for this in 739b2e8.